app/controllers/application_controller.rb

```diff
@@ -15,7 +15,41 @@ def current_order
     if session[:order_id]
       Order.find(session[:order_id])
     else
-      Order.new
+      new_order = Order.create
+      session[:order_id] = new_order.id
+      new_order
     end
   end
+
+  def authenticate_user
+    if session[:user_id]
+       # set current user object to @current_user object variable
+      @current_user = User.find(session[:user_id])
+      return true
+    else
+      redirect_to(login_path)
+      return false
+    end
+  end
+
+  def save_login_state
+    if session[:user_id]
+      redirect_to user_path(@user.id)
+      return false
+    else
+      return true
+    end
+  end
+
+  def error_messages(instance)
+    error_array = []
+    instance.errors.messages.each do |key, value|
+      error_array += [[key.to_s.capitalize, value.first]]
+    end
+    error_array
+  end
+
+  def category_exists_for_product?(prod_id, cat_id)
```

Comment:
Given that this is only called in the `ProductCategoriesController`, I'd say this method should live there as a `private` method.